File: bindings/wasm/src/ts/prim.ts
// Copyright (c) 2026 Capgemini Engineering Research and Development.
//
// This file is part of OCCT-Light software library.
//
// This library is free software; you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License version 3 as published
// by the Free Software Foundation, with an option to use any later version.
// Consult the file LICENSE_AGPL_30.txt included in OCCT-Light distribution
// for complete text of the license and disclaimer of any warranty.
//
// Alternatively, this file may be used under the terms of a commercial
// license or contractual agreement.
//
// SPDX-License-Identifier: AGPL-3.0-or-later

import type { LoadedOcctl } from "./abi.js";
import { check } from "./abi.js";
import type { Graph } from "./topo.js";
import type { Uid } from "./ids.js";
import { NodeId } from "./ids.js";
import { UnsupportedError } from "./errors.js";
import { FUNCTION_NAMES_PRIM, bindRawPrim } from "./generated/prim.js";
import { STRUCT_LAYOUTS, mallocStruct, setField } from "./generated/_layouts.js";

export { FUNCTION_NAMES_PRIM };

/** Options for makeBox. The binding sets struct_version itself. */
export interface BoxInfo {
  dx: number;
  dy: number;
  dz: number;
  /** Optional local frame. Defaults to XOY. */
  placement?: {
    location?: { x: number; y: number; z: number };
    x_dir?: { x: number; y: number; z: number };
    x_dir_ref?: { x: number; y: number; z: number };
    normal?: { x: number; y: number; z: number };
    xDirection?: { x: number; y: number; z: number };
  };
}

/** Options for makeSphere. The binding sets struct_version itself. */
export interface SphereInfo {
  radius: number;
  angle1?: number;
  angle2?: number;
  angle?: number;
}

/** Options for makeCylinder. The binding sets struct_version itself. */
export interface CylinderInfo {
  radius: number;
  height: number;
  angle?: number;
}

/** Options for makeCone. The binding sets struct_version itself. */
export interface ConeInfo {
  r1: number;
  r2: number;
  height: number;
  angle?: number;
}

/** Options for makeTorus. The binding sets struct_version itself. */
export interface TorusInfo {
  r1: number;
  r2: number;
  angle1?: number;
  angle2?: number;
  angle?: number;
}

/** Options for makeWedge. The binding sets struct_version itself. */
export interface WedgeInfo {
  dx: number;
  dy: number;
  dz: number;
  ltx: number;
}

export class PrimModule {
  private readonly _raw: ReturnType<typeof bindRawPrim>;
  private readonly _occtl: LoadedOcctl;
  constructor(occtl: LoadedOcctl) {
    this._occtl = occtl;
    this._raw = bindRawPrim(occtl.raw);
  }
  get raw(): ReturnType<typeof bindRawPrim> { return this._raw; }

  /**
   * Build a box into `graph` and return its solid NodeId.
   *
   * Struct layout comes from the auto-generated STRUCT_LAYOUTS table
   * (computed with wasm32 alignment rules from the ABI catalogue).
   * The runtime `occtl_prim_box_info_init` initialiser sets defaults;
   * we overwrite dx/dy/dz using the generated offsets.
   */
  makeBox(graph: Graph, info: BoxInfo): NodeId {
    return this.makeSolid(graph, "occtl_prim_box_info_t", "occtl_prim_box_info_init", "occtl_prim_make_box", {
      dx: info.dx,
      dy: info.dy,
      dz: info.dz,
    }, info.placement);
  }

  /** Build a sphere into `graph` and return its solid NodeId. */
  makeSphere(graph: Graph, info: SphereInfo | number): NodeId {
    const options = typeof info === "number" ? { radius: info } : info;
    return this.makeSolid(graph, "occtl_prim_sphere_info_t", "occtl_prim_sphere_info_init", "occtl_prim_make_sphere", { ...options });
  }

  /** Build a cylinder into `graph` and return its solid NodeId. */
  makeCylinder(graph: Graph, info: CylinderInfo): NodeId {
    return this.makeSolid(graph, "occtl_prim_cylinder_info_t", "occtl_prim_cylinder_info_init", "occtl_prim_make_cylinder", { ...info });
  }

  /** Build a cone or truncated cone into `graph` and return its solid NodeId. */
  makeCone(graph: Graph, info: ConeInfo): NodeId {
    return this.makeSolid(graph, "occtl_prim_cone_info_t", "occtl_prim_cone_info_init", "occtl_prim_make_cone", { ...info });
  }

  /** Build a torus into `graph` and return its solid NodeId. */
  makeTorus(graph: Graph, info: TorusInfo): NodeId {
    return this.makeSolid(graph, "occtl_prim_torus_info_t", "occtl_prim_torus_info_init", "occtl_prim_make_torus", { ...info });
  }

  /** Build a wedge into `graph` and return its solid NodeId. */
  makeWedge(graph: Graph, info: WedgeInfo): NodeId {
    return this.makeSolid(graph, "occtl_prim_wedge_info_t", "occtl_prim_wedge_info_init", "occtl_prim_make_wedge", { ...info });
  }

  private makeSolid(
    graph: Graph,
    infoStruct: string,
    initSymbol: string,
    makeSymbol: string,
    fields: Record<string, number | undefined>,
    placement?: BoxInfo["placement"],
  ): NodeId {
    const mod = this._occtl.raw;
    const infoPtr = mallocStruct(infoStruct, mod);
    const idPtr = mod._malloc(8);
    try {
      // Zero the block then run the runtime initialiser.
      const layout = STRUCT_LAYOUTS[infoStruct];
      for (let i = 0; i < layout.size; i += 4) mod.setValue(infoPtr + i, 0, "i32");
      const initFn = (mod as unknown as Record<string, (p: number) => void>)[initSymbol];
      if (initFn) {
        initFn(infoPtr);
      } else {
        mod.setValue(infoPtr, 1, "i32");  // struct_version = 1
      }
      for (const [name, value] of Object.entries(fields)) {
        if (value !== undefined) setField(infoPtr, infoStruct, name, value, mod);
      }
      if (placement) this.writeAxis2Placement(infoPtr, placement);

      const fn = (mod as unknown as Record<string, (g: number, info: number, out: number) => number>)[makeSymbol];
      if (!fn) {
        throw new UnsupportedError(
          `Feature 'prim' is not available in this OCCT-Light build (missing ${makeSymbol})`,
          0n as Uid,
          0,
        );
      }
      const status = fn(graph.pointer, infoPtr, idPtr);
      check(mod, status);
      const lo = BigInt.asUintN(32, BigInt(mod.getValue(idPtr + 0, "i32") >>> 0));
      const hi = BigInt.asUintN(32, BigInt(mod.getValue(idPtr + 4, "i32") >>> 0));
      return NodeId.fromBits((hi << 32n) | lo);
    } finally {
      mod._free(infoPtr);
      mod._free(idPtr);
    }
  }

  private writeAxis2Placement(ptr: number, placement: NonNullable<BoxInfo["placement"]>): void {
    const mod = this._occtl.raw;
    const layout = STRUCT_LAYOUTS["occtl_prim_box_info_t"].fields.placement;
    const axis = STRUCT_LAYOUTS[layout.type];
    const base = ptr + layout.offset;
    const writeVec3 = (field: "location" | "x_dir" | "x_dir_ref", value?: { x: number; y: number; z: number }) => {
      if (!value) return;
      const offset = base + axis.fields[field].offset;
      mod.setValue(offset + 0, value.x, "double");
      mod.setValue(offset + 8, value.y, "double");
      mod.setValue(offset + 16, value.z, "double");
    };
    writeVec3("location", placement.location);
    writeVec3("x_dir", placement.x_dir ?? placement.normal);
    writeVec3("x_dir_ref", placement.x_dir_ref ?? placement.xDirection);
  }
}

File: bindings/rust/src/occtl/src/prim.rs
// Copyright (c) 2026 Capgemini Engineering Research and Development.
//
// This file is part of OCCT-Light software library.
//
// This library is free software; you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License version 3 as published
// by the Free Software Foundation, with an option to use any later version.
// Consult the file LICENSE_AGPL_30.txt included in OCCT-Light distribution
// for complete text of the license and disclaimer of any warranty.
//
// Alternatively, this file may be used under the terms of a commercial
// license or contractual agreement.
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use crate::error::{check, Result};
use crate::id::NodeId;
use crate::topo::Graph;
use occtl_sys as sys;

/// Convenience mirror of `occtl_prim_box_info_t`.
#[derive(Copy, Clone, Debug)]
pub struct BoxInfo {
    pub dx: f64,
    pub dy: f64,
    pub dz: f64,
    pub placement: sys::occtl_axis2_placement_t,
}

impl Default for BoxInfo {
    fn default() -> Self {
        let mut info: sys::occtl_prim_box_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_box_info_init(&mut info) };
        Self {
            dx: 0.0,
            dy: 0.0,
            dz: 0.0,
            placement: info.placement,
        }
    }
}

impl BoxInfo {
    pub fn new(dx: f64, dy: f64, dz: f64) -> Self {
        Self {
            dx,
            dy,
            dz,
            ..Self::default()
        }
    }
    pub fn with_placement(mut self, placement: sys::occtl_axis2_placement_t) -> Self {
        self.placement = placement;
        self
    }

    /// Project into the C ABI options struct.
    pub fn to_c(self) -> sys::occtl_prim_box_info_t {
        let mut info: sys::occtl_prim_box_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_box_info_init(&mut info) };
        info.dx = self.dx;
        info.dy = self.dy;
        info.dz = self.dz;
        info.placement = self.placement;
        info
    }
}

/// Convenience mirror of `occtl_prim_sphere_info_t`.
#[derive(Copy, Clone, Debug)]
pub struct SphereInfo {
    pub radius: f64,
    pub angle1: Option<f64>,
    pub angle2: Option<f64>,
    pub angle: Option<f64>,
    pub placement: sys::occtl_axis2_placement_t,
}

impl Default for SphereInfo {
    fn default() -> Self {
        let mut info: sys::occtl_prim_sphere_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_sphere_info_init(&mut info) };
        Self {
            radius: 0.0,
            angle1: None,
            angle2: None,
            angle: None,
            placement: info.placement,
        }
    }
}

impl SphereInfo {
    pub fn new(radius: f64) -> Self {
        Self {
            radius,
            ..Self::default()
        }
    }
    pub fn with_placement(mut self, placement: sys::occtl_axis2_placement_t) -> Self {
        self.placement = placement;
        self
    }

    /// Project into the C ABI options struct.
    pub fn to_c(self) -> sys::occtl_prim_sphere_info_t {
        let mut info: sys::occtl_prim_sphere_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_sphere_info_init(&mut info) };
        info.radius = self.radius;
        info.placement = self.placement;
        if let Some(v) = self.angle1 {
            info.angle1 = v;
        }
        if let Some(v) = self.angle2 {
            info.angle2 = v;
        }
        if let Some(v) = self.angle {
            info.angle = v;
        }
        info
    }
}

/// Convenience mirror of `occtl_prim_cylinder_info_t`.
#[derive(Copy, Clone, Debug)]
pub struct CylinderInfo {
    pub radius: f64,
    pub height: f64,
    pub angle: Option<f64>,
    pub placement: sys::occtl_axis2_placement_t,
}

impl Default for CylinderInfo {
    fn default() -> Self {
        let mut info: sys::occtl_prim_cylinder_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_cylinder_info_init(&mut info) };
        Self {
            radius: 0.0,
            height: 0.0,
            angle: None,
            placement: info.placement,
        }
    }
}

impl CylinderInfo {
    pub fn new(radius: f64, height: f64) -> Self {
        Self {
            radius,
            height,
            ..Self::default()
        }
    }
    pub fn with_placement(mut self, placement: sys::occtl_axis2_placement_t) -> Self {
        self.placement = placement;
        self
    }

    /// Project into the C ABI options struct.
    pub fn to_c(self) -> sys::occtl_prim_cylinder_info_t {
        let mut info: sys::occtl_prim_cylinder_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_cylinder_info_init(&mut info) };
        info.radius = self.radius;
        info.height = self.height;
        info.placement = self.placement;
        if let Some(v) = self.angle {
            info.angle = v;
        }
        info
    }
}

/// Convenience mirror of `occtl_prim_cone_info_t`.
#[derive(Copy, Clone, Debug)]
pub struct ConeInfo {
    pub r1: f64,
    pub r2: f64,
    pub height: f64,
    pub angle: Option<f64>,
    pub placement: sys::occtl_axis2_placement_t,
}

impl Default for ConeInfo {
    fn default() -> Self {
        let mut info: sys::occtl_prim_cone_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_cone_info_init(&mut info) };
        Self {
            r1: 0.0,
            r2: 0.0,
            height: 0.0,
            angle: None,
            placement: info.placement,
        }
    }
}

impl ConeInfo {
    pub fn new(r1: f64, r2: f64, height: f64) -> Self {
        Self {
            r1,
            r2,
            height,
            ..Self::default()
        }
    }
    pub fn with_placement(mut self, placement: sys::occtl_axis2_placement_t) -> Self {
        self.placement = placement;
        self
    }

    /// Project into the C ABI options struct.
    pub fn to_c(self) -> sys::occtl_prim_cone_info_t {
        let mut info: sys::occtl_prim_cone_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_cone_info_init(&mut info) };
        info.r1 = self.r1;
        info.r2 = self.r2;
        info.height = self.height;
        info.placement = self.placement;
        if let Some(v) = self.angle {
            info.angle = v;
        }
        info
    }
}

/// Convenience mirror of `occtl_prim_torus_info_t`.
#[derive(Copy, Clone, Debug)]
pub struct TorusInfo {
    pub r1: f64,
    pub r2: f64,
    pub angle1: Option<f64>,
    pub angle2: Option<f64>,
    pub angle: Option<f64>,
    pub placement: sys::occtl_axis2_placement_t,
}

impl Default for TorusInfo {
    fn default() -> Self {
        let mut info: sys::occtl_prim_torus_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_torus_info_init(&mut info) };
        Self {
            r1: 0.0,
            r2: 0.0,
            angle1: None,
            angle2: None,
            angle: None,
            placement: info.placement,
        }
    }
}

impl TorusInfo {
    pub fn new(r1: f64, r2: f64) -> Self {
        Self {
            r1,
            r2,
            ..Self::default()
        }
    }
    pub fn with_placement(mut self, placement: sys::occtl_axis2_placement_t) -> Self {
        self.placement = placement;
        self
    }

    /// Project into the C ABI options struct.
    pub fn to_c(self) -> sys::occtl_prim_torus_info_t {
        let mut info: sys::occtl_prim_torus_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_torus_info_init(&mut info) };
        info.r1 = self.r1;
        info.r2 = self.r2;
        info.placement = self.placement;
        if let Some(v) = self.angle1 {
            info.angle1 = v;
        }
        if let Some(v) = self.angle2 {
            info.angle2 = v;
        }
        if let Some(v) = self.angle {
            info.angle = v;
        }
        info
    }
}

/// Convenience mirror of `occtl_prim_wedge_info_t`.
#[derive(Copy, Clone, Debug)]
pub struct WedgeInfo {
    pub dx: f64,
    pub dy: f64,
    pub dz: f64,
    pub ltx: f64,
    pub placement: sys::occtl_axis2_placement_t,
}

impl Default for WedgeInfo {
    fn default() -> Self {
        let mut info: sys::occtl_prim_wedge_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_wedge_info_init(&mut info) };
        Self {
            dx: 0.0,
            dy: 0.0,
            dz: 0.0,
            ltx: 0.0,
            placement: info.placement,
        }
    }
}

impl WedgeInfo {
    pub fn new(dx: f64, dy: f64, dz: f64, ltx: f64) -> Self {
        Self {
            dx,
            dy,
            dz,
            ltx,
            ..Self::default()
        }
    }
    pub fn with_placement(mut self, placement: sys::occtl_axis2_placement_t) -> Self {
        self.placement = placement;
        self
    }

    /// Project into the C ABI options struct.
    pub fn to_c(self) -> sys::occtl_prim_wedge_info_t {
        let mut info: sys::occtl_prim_wedge_info_t = unsafe { std::mem::zeroed() };
        unsafe { sys::occtl_prim_wedge_info_init(&mut info) };
        info.dx = self.dx;
        info.dy = self.dy;
        info.dz = self.dz;
        info.ltx = self.ltx;
        info.placement = self.placement;
        info
    }
}

/// Builds an axis-aligned box and returns the new solid NodeId.
pub fn make_box(graph: &Graph, info: BoxInfo) -> Result<NodeId> {
    let raw = info.to_c();
    let mut id: sys::occtl_node_id_t = Default::default();
    check(unsafe { sys::occtl_prim_make_box(graph.as_ptr(), &raw, &mut id) })?;
    Ok(id.into())
}

/// Builds a sphere and returns the new solid NodeId.
pub fn make_sphere(graph: &Graph, info: SphereInfo) -> Result<NodeId> {
    let raw = info.to_c();
    let mut id: sys::occtl_node_id_t = Default::default();
    check(unsafe { sys::occtl_prim_make_sphere(graph.as_ptr(), &raw, &mut id) })?;
    Ok(id.into())
}

/// Builds a cylinder and returns the new solid NodeId.
pub fn make_cylinder(graph: &Graph, info: CylinderInfo) -> Result<NodeId> {
    let raw = info.to_c();
    let mut id: sys::occtl_node_id_t = Default::default();
    check(unsafe { sys::occtl_prim_make_cylinder(graph.as_ptr(), &raw, &mut id) })?;
    Ok(id.into())
}

/// Builds a cone or truncated cone and returns the new solid NodeId.
pub fn make_cone(graph: &Graph, info: ConeInfo) -> Result<NodeId> {
    let raw = info.to_c();
    let mut id: sys::occtl_node_id_t = Default::default();
    check(unsafe { sys::occtl_prim_make_cone(graph.as_ptr(), &raw, &mut id) })?;
    Ok(id.into())
}

/// Builds a torus and returns the new solid NodeId.
pub fn make_torus(graph: &Graph, info: TorusInfo) -> Result<NodeId> {
    let raw = info.to_c();
    let mut id: sys::occtl_node_id_t = Default::default();
    check(unsafe { sys::occtl_prim_make_torus(graph.as_ptr(), &raw, &mut id) })?;
    Ok(id.into())
}

/// Builds a wedge and returns the new solid NodeId.
pub fn make_wedge(graph: &Graph, info: WedgeInfo) -> Result<NodeId> {
    let raw = info.to_c();
    let mut id: sys::occtl_node_id_t = Default::default();
    check(unsafe { sys::occtl_prim_make_wedge(graph.as_ptr(), &raw, &mut id) })?;
    Ok(id.into())
}

/// Helper: builds an `occtl_axis2_placement_t` from origin (x,y,z) with the
/// canonical (Z up, X along +x) orientation.
pub fn placement_at(x: f64, y: f64, z: f64) -> sys::occtl_axis2_placement_t {
    sys::occtl_axis2_placement_t {
        location: sys::occtl_point3_t { x, y, z },
        x_dir: sys::occtl_direction3_t {
            x: 1.0,
            y: 0.0,
            z: 0.0,
        },
        x_dir_ref: sys::occtl_direction3_t {
            x: 0.0,
            y: 1.0,
            z: 0.0,
        },
    }
}
